packages/adblocker/src/engine/engine.ts

```diff
@@ -1174,16 +1178,32 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
 
     callerContext?: any | undefined;
   }): {
-    matches: {
-      filter: CosmeticFilter;
-      exception: CosmeticFilter | undefined;
-    }[];
+    matches: (
+      | {
+          filter: CosmeticFilter | undefined;
+          exception: CosmeticFilter | undefined;
+        }
+      | {
+          filter: undefined;
+          exception: NetworkFilter | undefined;
+        }
+    )[];
```

Comment:
This change is not backward compatible but suggested as we cannot bring all filters: see https://github.com/ghostery/adblocker/issues/5229#issuecomment-3371772383.

Reply:
cannot we do:
```
matches: {
  filter: CosmeticFilter | undefined;
  exception: CosmeticFilter | NetworkFilter | undefined;
}[];
```

Reply:
I prefer to have explicit typing as there's only two cases that generates `undefined` in `filter`. Also, I believe it's beneficial to allow TypeScript automatically infer if `exception` is `NetworkFilter` or `CosmeticFilter` by checking if `filter` is `undefined`.

Reply:
addressed by 3cf4a69b48547ef2f31f2a575bdb7d94427b3634